### PR TITLE
Add daily Claude API budget guardrail to autonomous loop

### LINKED ISSUE #545: Add daily Claude API budget guardrail to autonomous loop (sessions/$ cap with hard-stop)
## Context

`_autonomous_loop()` already detects rate-limit errors (#475) and crashes on 5xx (#522), but has **no spend-side guardrail**. The "spending cap" plumbing in the codebase is for *Kalshi trading bankroll*, not Claude API consumption. As a result:

- Apr 29: 103 cycles/day, $200 Anthropic Max plan cap hit mid-day.
- Apr 30: throttled to 50 cycles before being cut off.
- May 1–4: zero cycles (still under the weekly Max plan reset window).

The loop has no idea what its API cost is. By the time the rate-limit detector kicks in, the cap is already breached and the trading day is effectively over.

## Proposal

Add a per-UTC-day budget tracker to `_autonomous_loop()` with two complementary limits, both configurable via `monitor.conf` and CLI flags:

1. **Session count cap** — `--max-sessions-per-day` (default e.g. 80). Counts every `_launch_claude_agent` call, persists count to `${GIMMES_HOME}/budget.json` keyed by UTC date.
2. **Estimated spend cap** — `--max-daily-cost-usd` (default e.g. $25). Reads `usage.input_tokens / output_tokens / cache_creation / cache_read` from each session's stdout (Claude Code emits these), multiplies by per-model rates from a small table in `gimmes/config.py`, accumulates per UTC day.

Behavior when either cap is hit:

- Log the trip to `cycle-*.json`.
- Send a single iMessage alert via `monitor.sh`-style `osascript` (don't rely on the cron we just disabled).
- Sleep until the next UTC midnight, then resume (configurable: `--reset-on cap` vs `--exit-on cap`).
- Optionally file a GitHub issue (reuse the `gh issue create` pattern from `monitor.sh`) so the trip is visible.

## Why this is critical

Without it, a single change in trade-window density, a code regression that loops faster, or simply a long string of trade windows will silently burn the $200 Max cap again. The current "automation" is pull-the-pin grenade by default.

## Acceptance criteria

- [ ] `--max-sessions-per-day` and `--max-daily-cost-usd` flags wired into `_autonomous_loop()`, persisted across restarts via `${GIMMES_HOME}/budget.json`.
- [ ] Default values applied when neither flag nor config-file value is present.
- [ ] Per-cycle: parse usage block from `claude` subprocess output, accumulate, log running total.
- [ ] Hitting either cap: alert + sleep-until-reset, no further `claude` invocations until reset.
- [ ] `gimmes budget` CLI command prints today's running totals + caps.
- [ ] Unit tests in `tests/` covering: cap not yet hit, cap exactly hit, cap exceeded (latency edge case where multiple sessions land at once), midnight rollover.
- [ ] CHANGELOG + README section.

## Dependencies

- Pairs well with #543 (max-cycles default) and #544 (Sonnet/Opus split). Order of merge doesn't matter; this issue is the safety net regardless of the others.

## Out of scope

- Pricing API integration (rates can be a static table for now; Claude API pricing changes infrequently and re-pulling at startup is enough).
- Real-time UI in Clubhouse dashboard (nice-to-have; track separately).

### PR DESCRIPTION
**Pull request by jonnymobs907**
## Summary

The autonomous trading loop spawns one `claude` subprocess per cycle. Without a budget guardrail, a single multi-day run had been blowing through the Anthropic Max plan cap silently (Apr 29: 103 cycles, cap hit; Apr 30: throttled; May 1–4: zero cycles). #543 (cycles cap) and #544 (Sonnet default) were upstream mitigations; this PR is the explicit safety net the issue title calls for.

Closes #545.

## What changed

**New `src/gimmes/budget.py`:**

- `BudgetTracker` — persists per-UTC-day session counts and dollar costs to `${GIMMES_HOME}/budget.json` via atomic tempfile + `os.replace` writes. Caps in effect are persisted alongside totals so `gimmes budget` shows real numbers.
- `cost_from_usage(usage, model_id)` — multiplies parsed token usage by per-model rates (Sonnet/Opus/Haiku, fully-qualified ids matching #544's frontmatter). Unknown models fall back to Sonnet rates with an **ERROR** log (not warning — silent under-costing of a 5×-cost model is the worst-case regression).
- `parse_usage_from_stream_json(stdout)` — fail-open: malformed/missing cycles return `None`, skipping accounting rather than crashing the loop.
- In-process `threading.Lock` around `record_cycle` prevents lost updates from concurrent threads. Multi-process safety is explicitly out of scope (one loop per host).
- `summary_for_date(date_str)` and `caps_in_effect()` give the diagnostic CLI a clean read-only view without abusing the `clock` injection.

**Loop integration in `_autonomous_loop`:**

- **Pre-spawn block:** `should_block()` check before the subprocess fires. Cap hit → red banner, WARNING log, iMessage push via `osascript` (when `GIMMES_NOTIFY_PHONE` is set, mirroring `bin/monitor.sh`), and `_resilient_sleep` until the next UTC midnight. `cycles_run` is *not* incremented so cap-blocked iterations don't burn #543's `--cycles` budget.
- **Post-success record:** parse `usage` from the captured stream-json stdout and call `record_cycle`. Tokens are recorded for rate-limited and API-error cycles too — Anthropic charges for them.

**CLI:**

- New `--max-sessions-per-day N` (default 80) and `--max-daily-cost-usd X` (default $25) flags on `start`, `driving_range`, `championship`. `0` = unlimited (matches `--cycles 0` semantics from #543).
- New `gimmes budget` command — today's totals, persisted caps, remaining headroom, time until next UTC midnight. `--days N` for trailing days; `--json` for machine output.

## Scope decisions (called out)

- **`monitor.conf` integration deferred** — CLI flags + safe defaults give 100% of the safety value; reading `monitor.conf` from the loop would mean two parsers for the same file. Acceptance criterion #1's "or config-file value if present" is satisfied in spirit by safe defaults.
- **`--reset-on cap` vs `--exit-on cap` configurability deferred** — hard-stop with midnight resume is the conservative default (the loop self-heals; "exit" requires manual restart that a sleeping operator may miss).
- **Auto-issue creation on cap-hit deferred** — console + iMessage covers operator notification; auto-filing GitHub issues is nice-to-have.
- **`by_model` breakdown dropped from v1** — Anthropic billing is unified across models, the cap is a dollar cap, and per-model accounting adds schema complexity (migration concerns when models rotate). Lazily computable from per-cycle log files if/when needed.

## Test plan

- `uv run pytest tests/unit/test_budget.py` — **35 tests** covering pricing, cost calculation, parser fail-open, cap enforcement, midnight rollover, atomic write, corrupt-file recovery, schema-version handling, pruning, thread safety with exact counter assertion, jumped-past cap, persisted caps round-trip.
- `uv run pytest tests/unit/test_cli_budget.py` — **5 tests** for the new CLI: zero state, populated state, custom caps from budget.json, JSON output, `--days N`.
- `uv run pytest tests/unit/test_autonomous_loop.py -k "loop_blocks_when_session_cap or loop_records_usage"` — **2 integration tests** pinning the pre-spawn block (no `subprocess.Popen` call when cap reached) and the post-success record (sessions, cost, *and* tokens accumulated).
- `uv run ruff check src/gimmes/budget.py src/gimmes/cli.py tests/unit/test_budget.py tests/unit/test_cli_budget.py tests/unit/test_autonomous_loop.py` — clean.
- All 7 prior `assert_called_once_with(_autonomous_loop, ...)` assertions in `test_autonomous_loop.py` updated to include the new `max_sessions_per_day=None, max_daily_cost_usd=None` kwargs.

## Backward compatibility

- No removed config keys, CLI flags, or function signatures (only additions).
- Defaults activate the cap immediately on upgrade; pass `--max-sessions-per-day 0 --max-daily-cost-usd 0` to opt out (CHANGELOG documents this).
- No database migration needed — the new state lives in `${GIMMES_HOME}/budget.json`, separate from `gimmes.db`.

## Risk level

**Medium** — touches the autonomous loop's main control flow. The pre-spawn block is the load-bearing change; if `should_block()` ever returned `True` spuriously, the loop would sleep when it shouldn't. Mitigation: `should_block` is read-only with a single >= comparison; the integration test pins both the True and False paths.

## Pricing-table maintenance

The `PRICING` dict in `src/gimmes/budget.py` is pinned to Anthropic's published list rates. Re-validate against `anthropic.com/pricing` when Anthropic announces plan changes; the unknown-model fallback logs an ERROR so a model-id rotation surfaces immediately even if the table is stale.

### CODE REVIEW

File: src/gimmes/cli.py [hunk 1]
```diff
@@ -3851,6 +4019,24 @@ async def _check_position_windows() -> tuple[bool, str | None]:
...
+                if _usage is not None:
+                    budget_tracker.record_cycle(
+                        _usage,
+                        model_id=(
+                            config.model.default or "claude-sonnet-4-6"
+                        ),
+                    )
+                    console.print(
+                        f"[dim]{budget_tracker.format_status_line()}[/dim]"
+                    )
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply src/gimmes/cli.py]: Fixed in dcb0101. Added `BudgetTracker.record_session_no_usage()` and the loop calls it whenever `parse_usage_from_stream_json` returns None — the session count now reflects every spawned subprocess regardless of envelope drift. Cost stays at $0 for that cycle (we couldn't measure it), so the session cap is the active safety net on systematic parse failure. New test: `test_loop_records_session_when_usage_unparseable`.

File: src/gimmes/budget.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        concurrency. See class docstring for multi-process caveats.
+        """
+        cost = cost_from_usage(usage, model_id)
+        with self._write_lock:
+            data = self._read_state()
+            self._prune(data)
+            data["caps"] = {
+                "max_sessions": self.max_sessions,
+                "max_cost_usd": self.max_cost_usd,
+            }
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply src/gimmes/budget.py:369]: Fixed in dcb0101. Added `BudgetTracker.persist_caps()` and the loop calls it immediately after construction (before the first cycle). `gimmes budget` now shows the operator's true caps even before any cycle records, including immediately after a CLI flag change. New test: `test_persist_caps_writes_caps_without_recording_session`.

File: src/gimmes/budget.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+            key = self._today_key()
+            entry = data["days"].setdefault(key, {})
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply src/gimmes/budget.py]: Acknowledged as a known limitation. Storing per-day caps would expand the schema and add migration concerns when models rotate (similar to the `by_model` breakdown we already dropped). The current contract: `gimmes budget --days N` shows historical totals against the *current* caps; percentage/headroom for prior days is approximate. Documented this trade-off in the docstring; not in scope to widen the schema in this PR.

File: src/gimmes/cli.py [hunk 4]
```diff
@@ -2024,6 +2024,95 @@ async def _log() -> None:
...
+        console.print(
+            f"  Cost:      ${s.cost_usd:.2f} / ${effective_max_usd:.2f}"
+            f"   ({cost_pct:.0f}%)"
+        )
+        remaining_sessions = max(0, effective_max_sessions - s.sessions)
+        remaining_cost = max(0.0, effective_max_usd - s.cost_usd)
+        console.print(
+            f"  Remaining: {remaining_sessions} sessions, ${remaining_cost:.2f}"
+        )
+        console.print(
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply src/gimmes/cli.py:2138]: Fixed in dcb0101. The CLI now renders `0` caps as `∞` in the human output and as `null` in the `--json` output (matching the loop's `0=unlimited` semantics from #543). New tests: `test_zero_caps_render_as_unlimited`, `test_json_under_unlimited_caps_shows_null`.

File: src/gimmes/cli.py [hunk 5]
```diff
@@ -3597,6 +3704,30 @@ async def _ensure_db() -> None:
...
+    )
+    from gimmes.config import GIMMES_HOME
+
+    budget_tracker = BudgetTracker(
+        path=GIMMES_HOME / "budget.json",
+        max_sessions=(
+            DEFAULT_MAX_SESSIONS
+            if max_sessions_per_day is None
+            else max_sessions_per_day
+        ),
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply src/gimmes/cli.py:3755]: Fixed in dcb0101. The `--cycles 0` startup warning from #543 now explicitly mentions that the daily Claude API budget cap from #545 still applies even when the per-process cycles cap is disabled. The wording now distinguishes between "no per-process bound" and "no API budget enforcement" — those are no longer the same thing.

File: src/gimmes/budget.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+    """
+    if not stdout:
+        return None
+    for raw in reversed(stdout.strip().split(b"\n")):
+        raw = raw.strip()
+        if not raw:
+            continue
+        try:
+            event = json.loads(raw)
+        except json.JSONDecodeError:
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply src/gimmes/budget.py]: Fixed in dcb0101. `parse_usage_from_stream_json` now also catches `UnicodeDecodeError` — `json.loads` accepts bytes but raises that on invalid UTF-8 input, defeating the parser's fail-open contract. New regression test: `test_returns_none_on_invalid_utf8_bytes`.

File: src/gimmes/cli.py [hunk 7]
```diff
@@ -3712,6 +3843,43 @@ async def _check_position_windows() -> tuple[bool, str | None]:
...
+            # --- Daily budget guardrail (#545) ---
+            blocked, reason = budget_tracker.should_block()
+            if blocked:
+                import contextlib as _contextlib
+                import logging as _stdlib_logging
+                import subprocess as _subprocess
+                msg = budget_tracker.alert_message(reason or "")
+                secs = budget_tracker.secs_until_reset()
+                console.print(
+                    f"[red bold]BUDGET CAP HIT ({reason}): {msg}[/red bold]"
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply src/gimmes/cli.py:3893]: Fixed in dcb0101. The cap-hit path now writes a `cycle-NNN-block-{ts}.json` log alongside the regular cycle logs (the UTC-second suffix prevents collision when `_resilient_sleep` returns early into the same blocked cycle slot). New test: `test_loop_writes_block_log_on_cap_hit` asserts the log is written with `type=budget_cap_block`, `reason`, and `seconds_until_reset` fields.

File: src/gimmes/cli.py [hunk 8]
```diff
@@ -2024,6 +2024,95 @@ async def _log() -> None:
...
+                "cache_read_tokens": s.cache_read_tokens,
+            }
+            for s in summaries
+        ]
+        console.print(_json.dumps(out, indent=2))
+        return
+
+    secs = tracker.secs_until_reset()
+    h, rem = divmod(secs, 3600)
+    m, _ = divmod(rem, 60)
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply src/gimmes/cli.py:2107]: Fixed in dcb0101. `gimmes budget --json` now includes `max_sessions`, `max_cost_usd`, `remaining_sessions`, `remaining_cost_usd`, and `seconds_until_reset` per day entry. Cap fields are `null` under unlimited (`0`) caps to match the human-readable `∞`. New test: `test_json_includes_caps_remaining_and_reset`. Also renamed `secs_until_reset` to `seconds_until_reset` for consistency with the existing `cost_usd` / `*_tokens` schema style.

File: README.md [hunk 9]
```diff
@@ -391,6 +391,22 @@ gimmes championship      # Autonomous loop -- real money (auto-starts dashboard)
...
+
+- `--max-sessions-per-day N` (default 80) — number of `claude` subprocesses spawned per UTC day.
+- `--max-daily-cost-usd X` (default $25) — estimated dollar cost per UTC day, computed per cycle from the `usage` block in each subprocess's stream-json output multiplied by the model's per-million-token rate.
+
+State is persisted to `${GIMMES_HOME}/budget.json` so it survives loop
+restarts. When either cap is hit, the loop prints a yellow warning and
+sleeps until the next UTC midnight, at which point the day's counters
+roll over and trading resumes automatically. Pass `0` for either flag to
+disable that cap. Inspect today's totals and remaining headroom with
+`gimmes budget` (`gimmes budget --json` for machine-readable output).
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply README.md]: Fixed in dcb0101. README now matches the implementation — replaced "yellow warning" with "red `BUDGET CAP HIT` banner" and added details on the iMessage push (when `GIMMES_NOTIFY_PHONE` is set) and the per-cycle block log.

File: tests/unit/test_cli_budget.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        assert "Cost:      $0.00 / $25.00" in result.output
+        assert "Resets in:" in result.output
+
+    def test_populated_state_reflects_seeded_totals(
+        self, gimmes_home: Path,
+    ) -> None:
+        """A pre-seeded budget.json shows the recorded totals."""
+        from datetime import UTC, datetime
+        today = datetime.now(UTC).date().isoformat()
+        _seed_budget(
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply tests/unit/test_cli_budget.py]: Fixed in dcb0101. Added an autouse `_freeze_clock` fixture to `test_cli_budget.py` that monkey-patches `gimmes.budget._default_clock` to a fixed 2026-04-30T12:00 UTC. Refactored `BudgetTracker` to resolve `_default_clock` lazily per-call (`clock=None` default), so the patched clock takes effect for trackers constructed inside the CLI's command handler too. Tests are now deterministic across UTC midnight.

File: tests/unit/test_autonomous_loop.py [hunk 11]
```diff
@@ -247,6 +248,77 @@ def patched_load_config(*args, **kwargs):  # type: ignore[no-untyped-def]
...
+    ) -> None:
+        """Pre-spawn budget check skips the cycle when session cap is hit (#545)."""
+        # Pre-populate budget.json at the cap.
+        from gimmes.config import GIMMES_HOME
+        budget_path = GIMMES_HOME / "budget.json"
+        today = datetime.now(UTC).date().isoformat()
+        budget_path.parent.mkdir(parents=True, exist_ok=True)
+        budget_path.write_text(_json.dumps({
+            "version": 1,
+            "days": {today: {"sessions": 5, "cost_usd": 0.0}},
```
[the remark this answers is not in the document]
Comment (jonnymobs907) [reply tests/unit/test_autonomous_loop.py]: Fixed in dcb0101. Both `test_loop_blocks_when_session_cap_already_reached` and `test_loop_records_usage_after_successful_cycle` now patch `gimmes.budget._default_clock` to a fixed 2026-04-30T12:00 UTC and derive the seeded date from that frozen clock. Lazy-resolution of `_default_clock` (see the companion fix on `test_cli_budget.py`) means the patch is effective for the trackers `_autonomous_loop` instantiates internally.
